sentry-apollo-3/src/main/java/io/sentry/apollo3/SentryApollo3RequestComposer.kt

```diff
@@ -0,0 +1,39 @@
+package io.sentry.apollo3
+
+import com.apollographql.apollo3.api.ApolloRequest
+import com.apollographql.apollo3.api.CustomScalarAdapters
+import com.apollographql.apollo3.api.Mutation
+import com.apollographql.apollo3.api.Operation
+import com.apollographql.apollo3.api.Query
+import com.apollographql.apollo3.api.Subscription
+import com.apollographql.apollo3.api.http.DefaultHttpRequestComposer
+import com.apollographql.apollo3.api.http.HttpRequest
+import com.apollographql.apollo3.api.http.HttpRequestComposer
+import com.apollographql.apollo3.api.variables
+
+class SentryApollo3RequestComposer(url: String) : HttpRequestComposer {
+    private val defaultHttpRequestComposer = DefaultHttpRequestComposer(url)
+
+    override fun <D : Operation.Data> compose(apolloRequest: ApolloRequest<D>): HttpRequest {
+        val httpRequest = defaultHttpRequestComposer.compose(apolloRequest)
+        val builder = httpRequest.newBuilder()
+            .addHeader(SentryApollo3HttpInterceptor.SENTRY_APOLLO_3_OPERATION_TYPE, operationType(apolloRequest))
+            .addHeader(SentryApollo3HttpInterceptor.SENTRY_APOLLO_3_OPERATION_NAME, apolloRequest.operation.name())
+
+        apolloRequest.scalarAdapters?.let {
+            builder.addHeader(SentryApollo3HttpInterceptor.SENTRY_APOLLO_3_VARIABLES, apolloRequest.operation.variables(it).valueMap.toString())
```

Comment:
Will this blow up if there's long variables? Do we need to truncate / skip if too long?

Reply:
Hmm, good point... Don't think it would blow up because header sizes are usually limited by the server side. However, in order to keep our request the Sentry Server small we should truncate/skip.
Is that something we could/should use `SentryOptions#maxRequestBodySize` for?

Reply:
`maxRequestBodySize` is about the body only, headers are added as it is, IIRC this should be guarded by `defaultPii`

Reply:
@marandaneto Actually these headers are meant to be passed internally only in order to work around the need to deserialize.
They are added by the `SentryApollo3RequestComposer` and then read by the `SentryApollo3HttpInterceptor` and then removed. Only if people use the `RequestComposer` without the interceptor will the headers be sent over the wire. But then they don't get any use out of it as the `Interceptor` is the one to send info to our Backend.

I think Alex's question was about size limits for headers in general.

As for my idea with the `maxRequestBodySize`: Depending on the Apollo Operation the variables could get quite big and would therefore increase our payload sent to the Backend. Should we set a fixed sizeLimit or just send the variables as they are?

We thought about guarding with `defaultPii` but decided against it, correct @adinauer ?

Reply:
I'd just send the variables as they are.
We don't do any trimming on the client, the server drops if it exceeds and reports to the `client_reports` feature.
They can use the filtering callbacks to drop stuff in case they need to.

@lbloder using the request's header for this workaround could be problematic, in case this is a request for a 3rd party and somebody isn't using the interceptor, we can be leaking some PII data, I guess?
If there's no way around getting everything we need either with both interceptors, I'd rather go with the option that has more useful data and open an issue on the `apollo3` repo with a feature request explaining our use case.
Maybe @bruno-garcia has opinions here.

Reply:
> We thought about guarding with defaultPii but decided against it, correct @adinauer ?

Yes, was decided internally to not have this behind PII flag as devs can strip the variables in callbacks if they want to.

> @lbloder using the request's header for this workaround could be problematic, in case this is a request for a 3rd party and somebody isn't using the interceptor, we can be leaking some PII data, I guess?

We'd only be sending the data to the server where the variables are sent to anyways.

Reply:
@adinauer about the PII flags, headers usually are a source of tokens, etc, which is different than some other stuff such as navigation args, if you all have considered this use case and agreed with it, I'd double check tho, I recall using this PII guard for headers elsewhere and this was a must.

Reply:
@adinauer 

> We'd only be sending the data to the server where the variables are sent to anyways.

Headers are not compressed by default on normal SSL iirc, not sure if this is a concern.
Please ping @bruno-garcia if needed.

Reply:
We're not just taking all headers but only selected one's. In `SentryRequestResolver` for Spring we take all headers but exclude certain ones we deem sensitive.

https://github.com/getsentry/sentry-java/blob/413862e108edbf58285cf3b5c42de134c96a3bbe/sentry-spring/src/main/java/io/sentry/spring/SentryRequestResolver.java#L20-L21

Reply:
> Headers are not compressed by default on normal SSL iirc, not sure if this is a concern.

They should be encrypted if that's what you're saying. As for compression I don't know but it'd only happen if they misconfigure anyways.

Reply:
> Right now, we don't send the Request or its Headers to our Backend in the Interceptors, neither are we doing this in e.g. SentryOkHttpInterceptor nor in the SentryFeignClient if I'm not mistaken. @marandaneto You said this is in our spec? Does this apply to mobile as well? Should we open issues for the other integrations as well?

Sorry, just found your original comment.  

As for my workaround with using headers to pass data between `Interceptors`, I don't see a problem if our headers are leaked, except that the request is getting bigger, because the headers we add only contain info that is in the request body anyways.

In terms of sending the whole request to the server, then I would definitely do it like in the `SentryRequestResolver` that @adinauer linked

Reply:
If we are not sending headers in this PR to Sentry, no need to worry, but if we do, we should guard against defaultPii, See https://github.com/getsentry/sentry-dart/blob/d75a8c698a4942816336f3bae9a50c87fa6be0a8/dio/lib/src/dio_event_processor.dart#L117

Reply:
Made it easier to use with an extension function that adds both interceptors. Also this should make it harder to misconfigure,

The only headers we would send, and only if the `SentryApollo3HttpInterceptor` isn't added, are those that contain `variables` `operationName` and `operationType` which are all present in the request body anyways.

I would split sending the `request` and its `headers` to a separate issue for the integrations that are not sending those yet. I.e. The Apollo Integration v2, the OkHttp Integration, the FeignClient and maybe others, will look through them and create issues for those that are not sending the `request` and its `headers` to our backend.

Wdyt?

Reply:
We have https://github.com/getsentry/sentry-java/issues/2148 to track.